batch/batch/batch.py

```diff
@@ -839,15 +839,18 @@ def __init__(self, id, attributes, callback, userdata, user,
         await db.batch.update_record(self.id, cancelled=True, closed=True)
         self.cancelled = True
         self.closed = True
-        jobs = await self.get_jobs()
-        for j in jobs:
+        for j in await self.get_jobs():
             await j.cancel()
         log.info(f'batch {self.id} cancelled')
 
     async def close(self):
         await db.batch.update_record(self.id, closed=True)
         self.closed = True
 
+        for job in await self.get_jobs():
+            if job._state == 'Ready':
```

Comment:
Since start_job already checks for 'Ready', do we need the check here?

As an aside, it looks like we spawn 16 consumers, but in the same thread. Could we use a thread-safe Queue implementation, and spawn N threads? In particular, collections.dequeue appears to be thread safe for push/pop operations, and has performance very similar to lists (should be significantly better than Queue). This doesn't need to addressed for this PR, just something I thought could help future work. 

I found this exploration by David Beazley (creator of Curio) helpful: https://www.youtube.com/watch?v=MCs5OvhV9S4

Reply:
Also, since start_job awaits on the get, do we need to await on the put? Could use put_nowait, provided the queue size is infinite.

Still, to put less pressure on the queue, I would prefer pushing (eventually, not in this PR), a single collection of jobs.

Reply:
To recap: only question regarding the "Ready" check is for this PR, queue comment is related to broader batch
  - related to "Ready" question: do we potentially wait longer than necessary to start jobs that aren't ready but will be ready soon (Pending), or even never start these jobs (if `close` is never called again)? If so, it would be potentially slightly better to not check _state here at all, push the entire job list into the queue (easier to do as list in the future), and when the queue is ready to process that job in start_job, then the job is inspected for Ready state. This could protect against transitions happening in between `close` being called and `start_job` getting the item.